solara/server/starlette.py

```diff
@@ -121,9 +122,31 @@ async def process_messages_task(self):
             while len(self.to_send) > 0:
                 first = self.to_send.pop(0)
                 if isinstance(first, bytes):
-                    await self.ws.send_bytes(first)
+                    await self._send_bytes_exc(first)
                 else:
-                    await self.ws.send_text(first)
+                    await self._send_text_exc(first)
+
+    async def _send_bytes_exc(self, data: bytes):
+        # make sures we catch the starlette/websockets specific exception
+        # and re-raise it as a websocket.WebSocketDisconnect
+        try:
+            await self.ws.send_bytes(data)
+        except websockets.exceptions.ConnectionClosed as e:
+            raise websocket.WebSocketDisconnect() from e
+        except RuntimeError as e:
+            # starlette throws a RuntimeError once you call send after the connection is closed
+            raise websocket.WebSocketDisconnect() from e
```

Comment:
Should this be implemented as a decorator / context manager to not define two almost identical functions?

Reply:
As discussed in private, it's DRY (don't repeat yourself) vs YAGNI (you ain't gonna need it). It would  make it more difficult to read, vs it makes it more difficult to maintain.
In this particular case, we probably should have overwritten the `send(..)` method and should have pushed the branching down.